lib/at_monitor/at_monitor.c

```diff
@@ -35,26 +53,22 @@ void at_monitor_dispatch(const char *notif)
 
 	__ASSERT_NO_MSG(notif != NULL);
 
-	/* Dispatch to monitors in ISR, if any.
-	 * There might be non-ISR monitors that are interested
-	 * in the same notification, so copy it to the heap afterwards regardless.
-	 */
-	STRUCT_SECTION_FOREACH(at_monitor_isr_entry, e) {
-		if (!e->paused && (e->filter == ANY || strstr(notif, e->filter))) {
-			LOG_DBG("Dispatching to %p (ISR)", e->handler);
-			e->handler(notif);
-		}
-	}
-
 	monitored = false;
 	STRUCT_SECTION_FOREACH(at_monitor_entry, e) {
-		if (!e->paused && (e->filter == ANY || strstr(notif, e->filter))) {
-			monitored = true;
-			break;
+		if (!is_paused(e) && has_match(e, notif)) {
+			if (!is_direct(e)) {
+				/* Copy and schedule work-queue task */
+				monitored = true;
+				break;
```

Comment:
Should this `break` be here? The code might break out of the loop while there are still direct monitors in the list which should be called, right?

Reply:
You are correct, we should keep looping to dispatch this to direct monitors. Also, it the workqueue task we should check whether the monitor is direct or not, in which case we have already dispatched the notification to it and should not do it again. I have added that check now